File: client/src/models/issue/issue.ts
import { z } from 'zod';
import { IssueCommentSchema } from '@/models/comment/issueComment';
import { IssueIdSchema, ProjectIdSchema, UserIdSchema } from '@/models/ids';
import { PermissionSchema } from '@/models/permission/permission';
import { optionalDate } from '@/models/primitives';
import { IssuePrioritySchema } from './issuePriority';
import { IssueProgressSchema } from './issueProgress';
import { IssueTypeSchema } from './issueType';

export const IssueSchema = z.object({
  id: IssueIdSchema,
  projectId: ProjectIdSchema,
  type: IssueTypeSchema,
  progress: IssueProgressSchema,
  priority: IssuePrioritySchema,
  permission: PermissionSchema.nullish().transform((v) => v ?? undefined),
  summary: z.string(),
  description: z
    .string()
    .nullish()
    .transform((v) => v ?? ''),
  createdBy: z.string(),
  responsibleBy: z
    .array(UserIdSchema)
    .nullish()
    .transform((v) => v ?? []),
  creationTime: optionalDate,
  completionTime: optionalDate,
  commentPageCount: z.number().int().nonnegative().catch(0),
  comments: z
    .array(IssueCommentSchema)
    .nullish()
    .transform((v) => v ?? [])
});

export type Issue = z.infer<typeof IssueSchema>;

File: server/src/IssueTracker.Domain/Models/Issue.cs
using IssueTracker.Domain.Models.Enums;

namespace IssueTracker.Domain.Models;

public record Issue
{
    public string Id { get; set; }
    public string Summary { get; set; }
    public string Description { get; set; }
    public IssueType Type { get; set; } = IssueType.Bug;
    public IssueProgress Progress { get; set; } = IssueProgress.OnHold;
    public IssuePriority Priority { get; set; } = IssuePriority.Low;
    public string CreatedBy { get; set; }
    public DateTime CreationTime { get; set; }
    public DateTime? CompletionTime { get; set; }
    public string ProjectId { get; set; }
    public Project Project { get; set; }
    public IList<Permission> Permissions { get; set; } = new List<Permission>();
    public IEnumerable<Comment> Comments { get; set; } = new List<Comment>();
}
